File: solutions/topological-sort/javascript/0630-course.js
/**
### INTUITION:
[This problem requires understanding of topological sort concepts. The key insight is to identify the optimal approach for this specific scenario.]

### APPROACH:
1. **Analyze the problem**: Understand the input constraints and expected output
2. **Choose the right technique**: Apply topological sort methodology
3. **Implement efficiently**: Focus on optimal time and space complexity
4. **Handle edge cases**: Consider boundary conditions and special cases

### WHY THIS WORKS:
- This ensures that the solution leverages topological sort principles
- This ensures that time complexity is optimized for the given constraints
- This ensures that space complexity is minimized where possible

### EXAMPLE WALKTHROUGH:
Input:
```
courses = [[100,200],[200,1300],[1000,1250],[2000,3200]]
```

Step 1: Sort by end time
sorted = [[200,1300],[1000,1250],[2000,3200],[100,200]]
Step 2: Greedy selection
Take course ending at 1300
Take course ending at 1250 (can't, conflicts)
Take course ending at 3200
...

Output:
```
3 (max courses)
```

### TIME COMPLEXITY:
O(n)**
- Single pass through input

### SPACE COMPLEXITY:
O(1)**
- Constant extra space

### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

</details>

*/

/**
 * Main solution for Problem 630: Course Schedule III
 *
 * @param {number[][]} courses - Array of [duration, lastDay] pairs
 * @return {number} - Maximum number of courses that can be taken
 *
 * Time Complexity: O(n log n) for sorting and priority queue operations
 * Space Complexity: O(n) for the priority queue
 */
function solve(courses) {
  // Sort courses by their end day (deadline)
  courses.sort((a, b) => a[1] - b[1]);

  // Max heap to store durations (using negative values for max behavior)
  const maxHeap = [];
  let currentTime = 0;

  // Helper functions for max heap
  const heapPush = (val) => {
    maxHeap.push(val);
    let idx = maxHeap.length - 1;
    while (idx > 0) {
      const parent = Math.floor((idx - 1) / 2);
      if (maxHeap[parent] >= maxHeap[idx]) break;
      [maxHeap[parent], maxHeap[idx]] = [maxHeap[idx], maxHeap[parent]];
      idx = parent;
    }
  };

  const heapPop = () => {
    if (maxHeap.length === 0) return null;
    if (maxHeap.length === 1) return maxHeap.pop();

    const max = maxHeap[0];
    maxHeap[0] = maxHeap.pop();
    let idx = 0;

    while (true) {
      const left = 2 * idx + 1;
      const right = 2 * idx + 2;
      let largest = idx;

      if (left < maxHeap.length && maxHeap[left] > maxHeap[largest]) {
        largest = left;
      }
      if (right < maxHeap.length && maxHeap[right] > maxHeap[largest]) {
        largest = right;
      }
      if (largest === idx) break;

      [maxHeap[idx], maxHeap[largest]] = [maxHeap[largest], maxHeap[idx]];
      idx = largest;
    }

    return max;
  };

  // Greedy approach: take courses, replace with shorter ones if needed
  for (const [duration, lastDay] of courses) {
    currentTime += duration;
    heapPush(duration);

    // If we exceed the deadline, remove the longest course taken so far
    if (currentTime > lastDay) {
      currentTime -= heapPop();
    }
  }

  return maxHeap.length;
}

/**
 * Test cases for Problem 630: Course Schedule III
 */
function testSolution() {
  console.log("Testing 630. Course Schedule III");

  // Test case 1: Can take 3 courses
  const result1 = solve([
    [100, 200],
    [200, 1300],
    [1000, 1250],
    [2000, 3200],
  ]);
  const expected1 = 3;
  console.assert(
    result1 === expected1,
    `Test 1 failed: expected ${expected1}, got ${result1}`,
  );

  // Test case 2: Can only take 1 course
  const result2 = solve([[1, 2]]);
  const expected2 = 1;
  console.assert(
    result2 === expected2,
    `Test 2 failed: expected ${expected2}, got ${result2}`,
  );

  // Test case 3: Can take 2 courses
  const result3 = solve([
    [3, 2],
    [4, 3],
  ]);
  const expected3 = 0;
  console.assert(
    result3 === expected3,
    `Test 3 failed: expected ${expected3}, got ${result3}`,
  );

  // Test case 4: Multiple courses with same deadline
  const result4 = solve([
    [5, 5],
    [4, 6],
    [2, 6],
  ]);
  const expected4 = 2;
  console.assert(
    result4 === expected4,
    `Test 4 failed: expected ${expected4}, got ${result4}`,
  );

  // Test case 5: All courses can be taken
  const result5 = solve([
    [1, 2],
    [2, 4],
    [3, 7],
  ]);
  const expected5 = 3;
  console.assert(
    result5 === expected5,
    `Test 5 failed: expected ${expected5}, got ${result5}`,
  );

  console.log("All test cases passed for 630. Course Schedule III!");
}

/**
 * Example usage and demonstration
 */
function demonstrateSolution() {
  console.log("\n=== Problem 630. Course ===");
  console.log("Category: Topological Sort");
  console.log("Difficulty: Medium");
  console.log("");

  // Example demonstration would go here
  testSolution();
}

// Run tests if this file is executed directly
if (require.main === module) {
  demonstrateSolution();
}

// Export for use in other modules
module.exports = {
  solve,
  testSolution,
  demonstrateSolution,
};

/**
 * Additional Notes:
 * - This solution focuses on topological sort concepts
 * - Consider the trade-offs between time and space complexity
 * - Edge cases are crucial for robust solutions
 * - The approach can be adapted for similar problems in this category
 */

File: solutions/topological-sort/python/0630-course.py
"""
# Difficulty: Medium

# 0630. Course Schedule III

There are n different online courses numbered from 1 to n. You are given an array courses where courses[i] = [durationi, lastDayi] indicate that the ith course should be taken continuously for durationi days and must be finished before or on lastDayi.

You will start on the 1st day and you cannot take two or more courses simultaneously.

Return the maximum number of courses that you can take.

**Example:**

<dl class="example-details">
<dt>Input:</dt>
<dd>courses = [[100,200],[200,1300],[1000,1250],[2000,3200]]</dd>
<dt>Output:</dt>
<dd>3 (max courses)</dd>
<dt>Explanation:</dt>
<dd>Maximum courses you can take within time limit</dd>
</dl>

<details>
<summary><b>🔍 SOLUTION EXPLANATION</b></summary>
### METADATA:
**Techniques**: Hash Table Lookup, Hash Map Storage, Sorting
**Data Structures**: Hash Map, Hash Set, Array
**Patterns**: Greedy Algorithm
**Time Complexity**: O(n)
**Space Complexity**: O(1) - Constant extra space

### INTUITION:
The key insight is that greedy: sort courses by deadline. For each course, if time available, add to heap (track duration). If no time, compare with longest course taken. If current shorter, replace longest.

### APPROACH:
1. **Sort by end time**: Sort courses by their deadlines
2. **Initialize heap and time**: Create max heap, set current_time = 0
3. **Iterate courses**: For each course (duration, deadline)
4. **Add to schedule**: Push -duration to heap, add duration to current_time
5. **Check feasibility**: If current_time > deadline, remove longest course
6. **Continue processing**: Handle all courses
7. **Return result**: Return len(heap)

### WHY THIS WORKS:
- Sort courses by deadline: greedy choice of course order
- Use max heap to track course durations taken so far
- If current time + duration > deadline, remove longest course (heap top)
- Greedy: always try to fit course, remove longest if conflicts
- O(n log n) for sort + heap operations, O(n) space for heap

### EXAMPLE WALKTHROUGH:
Input:
```
courses = [[100,200],[200,1300],[1000,1250],[2000,3200]]
```

Step 1: Sort by end time
sorted = [[200,1300],[1000,1250],[2000,3200],[100,200]]
Step 2: Greedy selection
Take course ending at 1300
Take course ending at 1250 (can't, conflicts)
Take course ending at 3200
...

Output:
```
3 (max courses)
```

### TIME COMPLEXITY:
O(n)
- Single pass through input


### SPACE COMPLEXITY:
O(1)
- Constant extra space


### EDGE CASES:
- **Empty input**: Handle when input is empty
- **Single element**: Handle single-element inputs
- **Boundary values**: Handle minimum/maximum valid values

</details>
"""

import heapq

from typing import Any, List, Optional, Dict, Tuple


class Solution:
    def scheduleCourse(self, courses: List[List[int]]) -> int:
        """
        Finds the maximum number of courses that can be taken given their durations
        and deadlines.

        Args:
            courses: List of [duration, deadline] pairs for each course

        Returns:
            Maximum number of courses that can be taken

        Time Complexity: O(n log n) where n is the number of courses
        Space Complexity: O(n) for storing courses in the heap
        """
        # Sort courses by deadline to process earlier deadlines first
        courses.sort(key=lambda x: x[1])

        # Max heap to store course durations we've taken
        taken: list[Any] = []
        current_time = 0

        for duration, deadline in courses:
            # If we can take this course without exceeding its deadline
            if current_time + duration <= deadline:
                current_time += duration
                # Push negative duration for max heap (Python has min heap by default)
                heapq.heappush(taken, -duration)
            # If we can't take it directly, check if we can swap with a longer course
            elif taken and -taken[0] > duration:
                # Remove the longest course we've taken so far
                longest_duration = -heapq.heappop(taken)
                # Add the time difference after swapping
                current_time = current_time - longest_duration + duration
                # Add the new course
                heapq.heappush(taken, -duration)

        # Return the number of courses taken
        return len(taken)


def test_solution() -> None:
    """
    Test cases for the solution.
    """
    solution = Solution()

    # Test case 1: Example from problem
    result = solution.scheduleCourse([[100, 200], [200, 1300], [1000, 1250], [2000, 3200]])
    expected = 3
    assert result == expected, f"Expected expected, got result"

    # Test case 2: Simple case
    result = solution.scheduleCourse([[1, 2]])
    expected = 1
    assert result == expected, f"Expected expected, got result"

    # Test case 3: Multiple courses with tight deadlines
    result = solution.scheduleCourse([[3, 2], [4, 3]])
    expected = 0
    assert result == expected, f"Expected expected, got result"

    # Test case 4: Empty input
    result = solution.scheduleCourse([])
    expected = 0
    assert result == expected, f"Expected expected, got result"

    # Test case 5: Course duration equals deadline
    result = solution.scheduleCourse([[5, 5], [4, 6], [2, 6]])
    expected = 2
    assert result == expected, f"Expected expected, got result"

    print("All test cases passed!")


if __name__ == "__main__":
    test_solution()

    # Example usage
    solution = Solution()
    print(f"Solution for 630. Course")
